### PR TITLE
Build, test and ship with JDK 21

### PR DESCRIPTION
**Pull request by wendigo**
On top of the https://github.com/trinodb/trino-gateway/pull/224

Fixes https://github.com/trinodb/trino-gateway/issues/79

### CODE REVIEW

**Comment by mosabua** [comment]
We will discuss this as the next dev sync.

Probably also needs an update in the docker container build setup though.

**Comment by wendigo** [comment]
@mosabua it's already updated.

**Comment by wendigo** [comment]
It wouldn't work as the class file format have changed to 65. Docker build validates whether gateway starts.

File: .java-version [hunk 1]
```diff
@@ -1 +1 @@
-17.0.9
```
Comment (wendigo) [inline .java-version:1]: This updates the docker JVM version @mosabua

**Comment by wendigo** [comment]
And in the docker build logs:

```
#7 [jdk-download 2/2] RUN     set -xeuo pipefail &&     microdnf install -y tar gzip &&     echo "Downloading JDK from https://api.adoptium.net/v3/binary/version/jdk-21.0.1+12/linux/ppc64le/jdk/hotspot/normal/eclipse?project=jdk" &&     mkdir -p "/usr/lib/jvm/jdk-21.0" &&     curl -#LfS "https://api.adoptium.net/v3/binary/version/jdk-21.0.1+12/linux/ppc64le/jdk/hotspot/normal/eclipse?project=jdk" | tar -zx --strip 1 -C "/usr/lib/jvm/jdk-21.0"
```

File: .java-version (hunk 1 above)
Comment (mosabua) [reply .java-version:1]: oh .. cool

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -28,7 +28,7 @@ jobs:
       - name: Set up JDK 17
         uses: actions/setup-java@v3
         with:
-          java-version: '17'
+          java-version: '21'
```
Comment (ebyhr) [inline .github/workflows/ci.yml:31]: > - name: Set up JDK 17

Please update the name.

File: .java-version [hunk 3]
```diff
@@ -1 +1 @@
...
+21.0
```
Comment (ebyhr) [inline .java-version:1]: Please update development.md

https://github.com/trinodb/trino-gateway/blob/main/docs/development.md#build-requirements

**Review by ebyhr — APPROVED** [review]

File: .java-version (hunk 3 above)
Comment (Chaho12) [inline .java-version:1]: Yes this is much convenient without having patch version 👍

**Review by Chaho12 — APPROVED** [review]
LGTM 👍

**Comment by wendigo** [comment]
@mosabua ready to merge?

**Comment by mosabua** [comment]
We said we will hold for a few releases .. and just got 6 out. Maybe after 7 or 8 .. we did not agree to merge in todays dev sync so lets hold for now and put it on as discussion item for next dev sync

**Comment by willmostly** [comment]
I observed a SEGV during the test phase when using `OpenJDK 64-Bit Server VM Temurin-21.0.2+13 (build 21.0.2+13-LTS, mixed mode)` in combination with Docker Desktop on an M1 Mac. Using Temurin Java 21 with Orbstack or Temurin Java 17 with Docker Desktop built successfully. 

I don't think this should be a blocker, but it would be a good idea IMO for others to test.

In case it is interesting, the crash report included the below. I can share more if it is interesting to anyone
```
Crashed Thread:        0  Dispatch queue: com.apple.main-thread

Exception Type:        EXC_BAD_ACCESS (SIGSEGV)
Exception Codes:       KERN_PROTECTION_FAILURE at 0x000000016ef33c80
Exception Codes:       0x0000000000000002, 0x000000016ef33c80

Termination Reason:    Namespace SIGNAL, Code 11 Segmentation fault: 11
Terminating Process:   exc handler [794]

VM Region Info: 0x16ef33c80 is in 0x16b730000-0x16ef34000;  bytes after start: 58735744  bytes before end: 895
      REGION TYPE                    START - END         [ VSIZE] PRT/MAX SHRMOD  REGION DETAIL
      MALLOC_MEDIUM (reserved)    167800000-168000000    [ 8192K] rw-/rwx SM=NUL  ...(unallocated)
      GAP OF 0x3730000 BYTES
--->  STACK GUARD                 16b730000-16ef34000    [ 56.0M] ---/rwx SM=NUL  ... for thread 0
      Stack                       16ef34000-16f730000    [ 8176K] rw-/rwx SM=PRV  thread 0

Thread 0 crashed with ARM Thread State (64-bit):
    x0: 0x0000000000001268   x1: 0x00000001280574a0   x2: 0x000000000000dcd8   x3: 0x0000000000000e91
    x4: 0x00000001282a7d40   x5: 0x0000000000000088   x6: 0x0000000000005200   x7: 0x0000000000000000
    x8: 0x0000000000001d28   x9: 0x000000000000fa00  x10: 0x0000000128050000  x11: 0x000000000000007f
   x12: 0x0000000000000050  x13: 0x00000001590392f0  x14: 0x023e0001ea2481a1  x15: 0x00000001ea2481a0
   x16: 0xe26300018eaaef88  x17: 0x00000001007947a0  x18: 0x0000000000000000  x19: 0x000000016f72d0a0
   x20: 0x00000001280574a0  x21: 0x0000000000001268  x22: 0x0000000000000001  x23: 0x0000000000000000
   x24: 0x0000000000001d28  x25: 0xffffffffffffffff  x26: 0x00000000141300c9  x27: 0x0000600003c44190
   x28: 0x6d5900018fb6a390   fp: 0x000000016ef340e0   lr: 0xed0600010070a79c
    sp: 0x000000016ef33c90   pc: 0x000000018eaaef8c cpsr: 0x20001000
   far: 0x000000016ef33c80  esr: 0x92000047 (Data Abort) byte write Translation fault
```

**Comment by oneonestar** [comment]
Can't reproduce the problem using `OpenJDK 64-Bit Server VM Temurin-21.0.2+13 (build 21.0.2+13-LTS, mixed mode, sharing)` on `Ubuntu 22.04.4 LTS`.
```
➜  trino-gateway git:(serafin/jdk-21) $JAVA_HOME/bin/java -version
openjdk version "21.0.2" 2024-01-16 LTS
OpenJDK Runtime Environment Temurin-21.0.2+13 (build 21.0.2+13-LTS)
OpenJDK 64-Bit Server VM Temurin-21.0.2+13 (build 21.0.2+13-LTS, mixed mode, sharing)

➜  trino-gateway git:(serafin/jdk-21) $ ./mvnw clean test
...
[INFO] Results:
[INFO] 
[INFO] Tests run: 98, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for trino-gateway-parent 6-SNAPSHOT:
[INFO] 
[INFO] trino-gateway-parent ............................... SUCCESS [  1.613 s]
[INFO] gateway-ha ......................................... SUCCESS [01:20 min]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:22 min
[INFO] Finished at: 2024-03-09T17:54:23+09:00
[INFO] ------------------------------------------------------------------------

```

**Comment by mosabua** [comment]
Time to get this in! 💥

**Pull request merged**